codi_arduino2705.ino

```diff
@@ -1 +1,48 @@
+#include <SoftwareSerial.h>
+#include <TinyGPS++.h>
 
+TinyGPSPlus gps;
+SoftwareSerial gpsSerial(4, 3); // RX, TX per al GPS
+
+#include <SoftwareSerial.h>
+SoftwareSerial SerialBT(8, 9); // RX, TX per al mòdul Bluetooth
+
+float latitudWaypoint = 41.349389; // Latitud del waypoint
+float longitudWaypoint = 2.212583; // Longitud del waypoint
+float lastDistance = 0; // Distància anterior al waypoint
+
+void setup() {
+  Serial.begin(9600);
+  gpsSerial.begin(9600);
+  SerialBT.begin(9600); // Velocitat de comunicació del mòdul Bluetooth
+}
+
+void loop() {
+  while (gpsSerial.available() > 0) {
+    if (gps.encode(gpsSerial.read())) {
+      float currentLat = gps.location.lat();
+      float currentLon = gps.location.lng();
+      float distance = calcularDistancia(currentLat, currentLon, latitudWaypoint, longitudWaypoint);
+
+      if ((distance < 200 && lastDistance >= 200) || (distance >= 200 && lastDistance < 200) || (distance < 200 && millis() % 10000 == 0) || (distance >= 200 && millis() % 5000 == 0)) {
```

Comment:
La condició aquesta és molt llarga i costa d'entendre. Igual seria millor separar-la en vàries parts. Per altra banda, el 200 que estàs utilitzant 4 vegades el crearia for del loop com a constant. Per exemple:

```c
// Fora el loop
const int distanceThreshold = 200;
...

// Dins del loop
bool isCloseDistance = distance < distanceThreshold && lastDistance >= distanceThreshold;
bool wasCloseDistance = distance >= distanceThreshold && lastDistance < distanceThreshold;

if (isCloseDistance || wasCloseDistance) { ... }
```

El tema dels millis no l'acabo d'entendre... és per fer una vibració periòdica? Si és així s'hauria de fer diferent també. Digue'm què busques amb això i ho comentem.